### ISSUE TITLE
Exceeding host connection threshold when using table storage

### ISSUE THREAD

**Issue by vludax** [Azure/azure-functions-host#3814]
We are running into the the connection threshold limit when we attempt to update a few hundred records in an Azure Storage table. Could somebody please advise the best way to avoid hitting the limits? Awaiting each replace task before starting the next does seem to alleviate the issue, despite running considerably slower. 

- Function App version: 2.0
- Function App name: starfishimportergxdlh4wikzicy 

<details>
<summary>Source (simple reproduction)</summary>

```csharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Azure.WebJobs;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Table;

namespace StarFish.Configuration.StarCareImport.Functions
{
	public static class TimerFunctions
	{
		private const string TableName = "TestTable";
		private const string ConnectionStringSettingName = "StorageConnectionString";
		private const int RecordCount = 300;

		private static readonly TableStorage _storage = new TableStorage(new ConfigurationBuilder()
			.SetBasePath(Directory.GetCurrentDirectory())
			.AddJsonFile("local.settings.json", optional: true, reloadOnChange: true)
			.AddEnvironmentVariables()
			.Build());

		[FunctionName(nameof(TriggerStarCareImport))]
		public static Task TriggerStarCareImport(
			[TimerTrigger("*/20 * * * * *")]TimerInfo timer,
			ExecutionContext ctx,
			ILogger log)
		{
			return Task.WhenAll(Enumerable.Range(0, RecordCount)
				.Select(i => _storage.ReplaceAsync(i, log)));
		}

		public class TableStorage
		{
			private readonly Lazy<Task<CloudTable>> _table;

			public TableStorage(IConfigurationRoot cfg)
			{
				string saConStr = cfg.GetValue<string>(ConnectionStringSettingName);
				CloudTableClient tc = CloudStorageAccount.Parse(saConStr).CreateCloudTableClient();
				_table = new Lazy<Task<CloudTable>>(async () =>
				{
					CloudTable table = tc.GetTableReference(TableName);
					await table.CreateIfNotExistsAsync().ConfigureAwait(false);
					return table;
				});
			}

			public async Task ReplaceAsync(int i, ILogger log)
			{
				try
				{
					log.LogDebug("Replacing {i}", i);
					CloudTable table = await _table.Value;

					TableResult result = await table.ExecuteAsync(TableOperation.InsertOrReplace(new DynamicTableEntity(
						partitionKey: i.ToString(),
						rowKey: i.ToString(),
						etag: "*",
						properties: new Dictionary<string, EntityProperty>
						{
							["something"] = new EntityProperty(Guid.NewGuid())
						})));

					log.LogInformation("Replaced {i}", i);
				}
				catch (Exception e)
				{
					log.LogError(e, "Failed replacing {i}", i);
					throw;
				}
			}
		}
	}
}
```
</details>

**Comment by fabiocav** [comment]
@brettsam can you please link to the issue we have tracking this work?

**Comment by dallancarr** [comment]
Until this get investigated maybe await on the `ReplaceAsync` instead with an added delay between invocations for a small throttle.

**Comment by brettsam** [comment]
This is on the consumption plan, correct? Could you give this a try with the latest Functions release [2.0.12210.0](https://github.com/Azure/azure-functions-host/releases/tag/v2.0.12210)? We added a workaround that should help here.

**Comment by dallancarr** [comment]
So MaxConnectionPerServer was defaulting? Interesting to know what the fix was. 😁

**Comment by brettsam** [comment]
Copied from another issue ([here](https://github.com/Azure/azure-functions-host/issues/3781#issuecomment-444157388)): 

> A little background -- .NET Core made the ServicePointManager a no-op when it comes to HttpClient usage. We used this in Functions v1 to keep a constraint on the number of connections a single HttpClient could create to a single host (using ServicePointManager.DefaultConnectionLimit). This now does nothing in .NET Core. This means that the various SDKs now need to give us a way to control this by passing an HttpClientHandler into them (or exposing some other means of controlling MaxConnectionsPerServer).
> 
> For Storage, specifically, we implemented a private-reflection workaround to change this setting back to 50, which is what it was in Functions v1: #3749. Our testing has shown this to work, but I'm trying to find all issues filed that look like this is the cause and see if we're missing any scenarios.

**Comment by brettsam** [comment]
@vludax -- what is your timer trigger doing? I wouldn't expect a timer to eat up that many connections by itself. It's still possible to hold onto connections by creating new HttpClients (rather than sharing a static one) or connecting to many different hosts from within a function. If you'd like to [share your function name](https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately), I can take a look from our end and see if any logs jump out at me.

Another potential issue is that if you're using HttpClient directly from within your function, you need to set `MaxConnectionsPerServer` yourself with something like:

```
private static readonly HttpClient Client = new HttpClient(new HttpClientHandler
{
    MaxConnectionsPerServer = 50
});
```

**Comment by vludax** [comment]
Thanks for looking into this for us @brettsam!

The timer-triggered function starts a new durable function, which then uses three singleton `CloudTableClient`s to insert/replace hundreds of rows. We're not manually instantiating `HttpClient`. The function name is starfishimportergxdlh4wikzicy.

**Comment by vludax** [comment]
Thanks @brettsam! There were indeed some records with `itemCount` > 1. We've added the settings you suggested to host.json, and will see how that goes.

We noticed that the only errors we're getting now are 404's from durable functions (/DurableFunctionsHubInstances endpoint), which we assume is unrelated. 

We'll will keep inserting records sequentially for now (this opens very few connections), but may attempt to parallelise it again at a later date.
